aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,630 @@
+            perm,
+            |stored_permission| {
+                increase_capacity(stored_permission, capacity);
+            },
+            StoredPermission::Capacity(capacity),
+        )
+    }
+
+    public(package) fun check_permission_exists<PermKey: copy + drop + store>(
+        s: &signer, perm: PermKey
+    ): bool acquires PermissionStorage {
+        check_permission_capacity_above(s, 0, perm)
+    }
+
+    public(package) fun check_permission_capacity_above<PermKey: copy + drop + store>(
+        s: &signer, threshold: u256, perm: PermKey
+    ): bool acquires PermissionStorage {
+        if (!is_permissioned_signer(s)) {
+            // master signer has all permissions
+            return true
+        };
+        map_or(
+            s,
+            perm,
+            |stored_permission| {
+                is_above(stored_permission, threshold)
+            },
+            false,
+        )
+    }
+
+    public(package) fun check_permission_consume<PermKey: copy + drop + store>(
+        s: &signer, threshold: u256, perm: PermKey
+    ): bool acquires PermissionStorage {
+        if (!is_permissioned_signer(s)) {
+            // master signer has all permissions
+            return true
+        };
+        map_or(
+            s,
+            perm,
+            |stored_permission| {
+                 consume_capacity(stored_permission, threshold)
+            },
+            false,
+        )
+    }
+
+    public(package) fun capacity<PermKey: copy + drop + store>(
+        s: &signer, perm: PermKey
+    ): Option<u256> acquires PermissionStorage {
+        if (!is_permissioned_signer(s)) {
+            return option::some(U256_MAX)
+        };
+        map_or(
+            s,
+            perm,
+            |stored_permission: &mut StoredPermission| {
+                option::some(match (stored_permission) {
+                    StoredPermission::Capacity(capacity) => *capacity,
+                    StoredPermission::Unlimited => U256_MAX,
+                })
+            },
+            option::none(),
+        )
+    }
+
+    public(package) fun revoke_permission<PermKey: copy + drop + store>(
+        permissioned: &signer, perm: PermKey
+    ) acquires PermissionStorage {
+        if (!is_permissioned_signer(permissioned)) {
+            // Master signer has no permissions associated with it.
+            return
+        };
+        let addr = permission_address(permissioned);
+        if (!exists<PermissionStorage>(addr)) { return };
+        let perm_storage = &mut PermissionStorage[addr].perms;
+        let key = copyable_any::pack(perm);
+        if (perm_storage.contains(&key)) {
+            perm_storage.remove(&key);
+        }
+    }
+
+    // =====================================================================================================
+    // Native Functions
+    ///
+    /// Check whether this is a permissioned signer.
+    public native fun is_permissioned_signer(s: &signer): bool;
```

Comment:
this cannot be public in a first release.

Reply:
I think this needs to be public so that 3rd party can at least abort on permissioned signer if needed

Reply:
@runtian-zhou is that needed for first release?

this is not backward compatible - you cannot add new public native function in one release.

if you think this is critical , you need to make this not be native, and have is_permissioned_signer_impl be private and native (and you can register both in rust code, so that you can later remove is_permissioned_signer_impl and change is_permissioned_signer to native

Reply:
https://aptos-org.slack.com/archives/C045G5Q0V8W/p1729612895923779?thread_ts=1729532945.900979&cid=C045G5Q0V8W

Reply:
Updated the natives.